cache/cache.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package cache provides a simple caching mechanism
+// that limits the age of cache entries and tries to avoid large
+// repopulation events by staggering refresh times.
+package cache
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"gopkg.in/errgo.v1"
+)
+
+type entry struct {
+	value  interface{}
+	expire time.Time
+}
+
+// Key represents a cache key. It must be a comparable type.
+type Key interface{}
+
+// Cache holds a time-limited cache of values for arbitrary keys.
+type Cache struct {
+	maxAge time.Duration
+
+	// mu guards the fields below it.
+	mu sync.Mutex
+
+	// expire holds when the cache is due to expire.
+	expire time.Time
+	// We hold two maps so that can avoid scanning through all the
+	// items in the cache when the cache needs to be refreshed.
+	// Instead, we move items from old to new when they're accessed
+	// and throw away the old map at refresh time.
+	old, new map[Key]entry
+}
+
+// New returns a new Cache that will cache items for
+// at most maxAge.
+func New(maxAge time.Duration) *Cache {
+	// A maxAge is < 2ns then the expiry code will panic because the
+	// actual expiry time will be maxAge - a random value in the
+	// interval [0. maxAge/2). If maxAge is < 2ns then this requires
+	// a random interval in [0, 0) which causes a panic.
+	if maxAge < 2*time.Nanosecond {
+		maxAge = 2 * time.Nanosecond
+	}
+	// The returned cache will have a zero-valued expire
+	// time, so will expire immediately, causing the new
+	// map to be created.
+	return &Cache{
+		maxAge: maxAge,
+	}
+}
+
+// Len returns the total number of cached entries.
+func (c *Cache) Len() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return len(c.old) + len(c.new)
+}
+
+// Evict removes the entry with the given key from the cache if present.
+func (c *Cache) Evict(key Key) {
```

Comment:
excellent point. I've added a test, checked that it failed, then fixed the logic.